File: karaoke/infra/scanner.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import asyncio
import os
import time
from dataclasses import dataclass, field
from typing import Dict, Iterable, List, Optional, Set, Tuple

from karaoke.infra.media import probe_video_playable
from karaoke.infra.models import Song
from settings import (
    DEFAULT_DUPLICATE_POLICY,
    FFPROBE_ON_IMPORT,
    KEEP_DIR_NAME,
    OVERRIDE_DIR_NAME,
    SCAN_VIDEO_EXTS,
    logger,
)

_VIDEO_EXTS = frozenset(SCAN_VIDEO_EXTS)
_SKIP_DIR_NAMES = frozenset({KEEP_DIR_NAME, OVERRIDE_DIR_NAME})
_BULK_BATCH = 500
_PREVIEW_LIMIT = 200
_UPDATE_FIELDS = (
    'display_name', 'source_path', 'source_rel', 'source_origin',
    'is_playable', 'scan_root', 'media_kind', 'playback_mode',
    'playback_source', 'can_queue',
)


@dataclass
class ScanStats:
    added: int = 0
    skipped: int = 0
    renamed: int = 0
    invalid: int = 0
    preview: List[dict] = field(default_factory=list)

    def as_dict(self) -> dict:
        return {
            'added': self.added,
            'skipped': self.skipped,
            'renamed': self.renamed,
            'invalid': self.invalid,
        }


def _display_base(filename: str) -> str:
    return os.path.splitext(filename)[0]


def _make_unique_display_name(base: str, source_rel: str, used: Set[str]) -> tuple:
    candidate = base
    if candidate not in used:
        return candidate, False
    rel_suffix = source_rel.replace(os.sep, '/').strip('./')
    if rel_suffix and rel_suffix != '.':
        candidate = f"{base} ({rel_suffix})"
        if candidate not in used:
            return candidate, True
    index = 2
    while True:
        candidate = f"{base} ({index})"
        if candidate not in used:
            return candidate, True
        index += 1


def _append_preview(stats: ScanStats, item: dict) -> None:
    if len(stats.preview) < _PREVIEW_LIMIT:
        stats.preview.append(item)


def _iter_video_files(root: str) -> Iterable[Tuple[str, str]]:
    """用 os.scandir 深度遍历，仅 yield (绝对路径, 所在目录)。"""
    stack = [root]
    while stack:
        dirpath = stack.pop()
        try:
            with os.scandir(dirpath) as entries:
                subdirs: List[str] = []
                for entry in entries:
                    try:
                        if entry.is_dir(follow_symlinks=False):
                            if entry.name not in _SKIP_DIR_NAMES:
                                subdirs.append(entry.path)
                            continue
                        if not entry.is_file(follow_symlinks=False):
                            continue
                        name = entry.name
                        dot = name.rfind('.')
                        if dot <= 0:
                            continue
                        if name[dot + 1:].lower() in _VIDEO_EXTS:
                            yield entry.path, dirpath
                    except OSError:
                        continue
                stack.extend(reversed(subdirs))
        except OSError:
            continue


def _collect_scan(
    root: str,
    policy: str,
    do_probe: bool,
    dry_run: bool,
    existing_by_path: Dict[str, Song],
    existing_by_name: Dict[str, Song],
    used_names: Set[str],
) -> Tuple[ScanStats, List[Song], List[Song]]:
    stats = ScanStats()
    to_create: List[Song] = []
    to_update: Dict[int, Song] = {}
    t0 = time.monotonic()
    file_count = 0

    for abs_path, dirpath in _iter_video_files(root):
        file_count += 1
        filename = os.path.basename(abs_path)
        base_name = _display_base(filename)
        source_rel = os.path.relpath(dirpath, root)
        if source_rel == '.':
            source_rel = ''

        is_playable = True
        if do_probe:
            is_playable = probe_video_playable(abs_path)
            if not is_playable:
                stats.invalid += 1
                if dry_run:
                    _append_preview(stats, {'path': abs_path, 'action': 'invalid'})
                continue

        if abs_path in existing_by_path:
            if policy == 'overwrite':
                song = existing_by_path[abs_path]
                song.is_playable = is_playable
                song.scan_root = root
                song.source_rel = source_rel or None
                song.playback_mode = 'plain'
                song.playback_source = 'plain'
                song.can_queue = is_playable
                to_update[song.id] = song
                if dry_run:
                    _append_preview(stats, {'path': abs_path, 'action': 'update'})
                else:
                    stats.added += 1
            else:
                stats.skipped += 1
                if dry_run:
                    _append_preview(stats, {'path': abs_path, 'action': 'skip'})
            continue

        display_name = base_name
        renamed = False
        existing_same_name = existing_by_name.get(display_name)
        if existing_same_name and existing_same_name.source_path != abs_path:
            if policy == 'skip':
                stats.skipped += 1
                if dry_run:
                    _append_preview(stats, {'path': abs_path, 'action': 'skip'})
                continue
            if policy == 'overwrite':
                if not dry_run:
                    existing_same_name.source_path = abs_path
                    existing_same_name.source_rel = source_rel or None
                    existing_same_name.is_playable = is_playable
                    existing_same_name.scan_root = root
                    existing_same_name.source_origin = 'scan'
                    existing_same_name.playback_mode = 'plain'
                    existing_same_name.playback_source = 'plain'
                    existing_same_name.can_queue = is_playable
                    to_update[existing_same_name.id] = existing_same_name
                    existing_by_path[abs_path] = existing_same_name
                stats.added += 1
                if dry_run:
                    _append_preview(stats, {'path': abs_path, 'action': 'overwrite'})
                continue
            display_name, renamed = _make_unique_display_name(base_name, source_rel, used_names)
            if renamed:
                stats.renamed += 1

        used_names.add(display_name)
        if dry_run:
            _append_preview(stats, {
                'path': abs_path,
                'action': 'rename' if renamed else 'add',
                'display_name': display_name,
            })
            stats.added += 1
            continue

        song = Song(
            display_name=display_name,
            source_path=abs_path,
            source_origin='scan',
            source_rel=source_rel or None,
            media_kind='video',
            playback_mode='plain',
            playback_source='plain',
            can_queue=is_playable,
            is_playable=is_playable,
            scan_root=root,
        )
        to_create.append(song)
        existing_by_path[abs_path] = song
        existing_by_name[display_name] = song
        stats.added += 1

    elapsed = time.monotonic() - t0
    logger.info(
        'scan collect: files=%s added=%s skipped=%s invalid=%s probe=%s elapsed=%.2fs',
        file_count, stats.added, stats.skipped, stats.invalid, do_probe, elapsed,
    )
    return stats, to_create, list(to_update.values())


async def scan_root(
    root: str,
    duplicate_policy: Optional[str] = None,
    validate: Optional[bool] = None,
    dry_run: bool = False,
) -> ScanStats:
    root = os.path.abspath(root)
    if not os.path.isdir(root):
        raise FileNotFoundError(root)

    policy = duplicate_policy or DEFAULT_DUPLICATE_POLICY
    do_probe = FFPROBE_ON_IMPORT if validate is None else validate

    songs = await Song.all().only(
        'id', 'source_path', 'display_name', 'source_rel',
        'source_origin', 'is_playable', 'scan_root', 'media_kind',
        'playback_mode', 'audio_layout',
    )
    existing_by_path = {s.source_path: s for s in songs}
    existing_by_name = {s.display_name: s for s in songs}
    used_names = set(existing_by_name.keys())

    stats, to_create, to_update = await asyncio.to_thread(
        _collect_scan,
        root,
        policy,
        do_probe,
        dry_run,
        existing_by_path,
        existing_by_name,
        used_names,
    )

    if not dry_run:
        t0 = time.monotonic()
        if to_create:
            await Song.bulk_create(to_create, batch_size=_BULK_BATCH)
        if to_update:
            await Song.bulk_update(to_update, fields=_UPDATE_FIELDS, batch_size=_BULK_BATCH)
        logger.info(
            'scan persist: create=%s update=%s elapsed=%.2fs',
            len(to_create), len(to_update), time.monotonic() - t0,
        )

    return stats

File: crates/infra/src/scanner.rs
//! 目录扫描导入。对应 Python `karaoke/infra/scanner.py`。

use crate::media::{probe_video_playable, MediaSettings};
use crate::models::NewSong;
use crate::repositories::SongRepository;
use futures::stream::{self, StreamExt};
use serde::Serialize;
use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use thiserror::Error;
use tracing::info;

const PREVIEW_LIMIT: usize = 200;

#[derive(Debug, Error)]
pub enum ScanError {
    #[error("扫描路径不存在或不可读: {0}")]
    RootNotFound(String),
    #[error("已有扫描任务正在运行")]
    Busy,
    #[error(transparent)]
    Db(#[from] sqlx::Error),
}

#[derive(Debug, Clone, Serialize)]
pub struct PreviewItem {
    pub path: String,
    pub action: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize)]
pub struct ScanStats {
    pub added: i64,
    pub skipped: i64,
    pub renamed: i64,
    pub invalid: i64,
    /// 落库后待后台校验的文件数（validate=true 时）。
    #[serde(default)]
    pub pending_validate: i64,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub preview: Vec<PreviewItem>,
}

impl ScanStats {
    fn push_preview(&mut self, item: PreviewItem) {
        if self.preview.len() < PREVIEW_LIMIT {
            self.preview.push(item);
        }
    }
}

#[derive(Debug, Clone, Default, Serialize)]
pub struct ScanValidateStatus {
    pub running: bool,
    pub total: i64,
    pub done: i64,
    pub invalid: i64,
}

/// 扫描校验后台进度（同一时间通常只有一个扫描校验在跑）。
#[derive(Clone, Default)]
pub struct ScanTaskManager {
    inner: Arc<Mutex<ScanValidateStatus>>,
}

impl ScanTaskManager {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn status(&self) -> ScanValidateStatus {
        self.inner.lock().unwrap().clone()
    }

    fn try_begin(&self, total: i64) -> bool {
        let mut g = self.inner.lock().unwrap();
        if g.running {
            return false;
        }
        *g = ScanValidateStatus {
            running: true,
            total,
            done: 0,
            invalid: 0,
        };
        true
    }

    fn tick(&self, invalid: bool) {
        let mut g = self.inner.lock().unwrap();
        g.done += 1;
        if invalid {
            g.invalid += 1;
        }
    }

    fn finish(&self) {
        let mut g = self.inner.lock().unwrap();
        g.running = false;
    }
}

pub struct ScanOptions {
    pub duplicate_policy: String,
    /// 同步校验（预览用）；正式扫描应设 false，校验改走后台。
    pub validate: bool,
    pub dry_run: bool,
}

fn walk_video_files(
    root: &Path,
    video_exts: &HashSet<String>,
    skip_dirs: &HashSet<String>,
) -> Vec<(PathBuf, PathBuf)> {
    let mut results = Vec::new();
    let mut stack = vec![root.to_path_buf()];
    while let Some(dir) = stack.pop() {
        let Ok(entries) = std::fs::read_dir(&dir) else {
            continue;
        };
        let mut subdirs = Vec::new();
        for entry in entries.flatten() {
            let path = entry.path();
            let Ok(file_type) = entry.file_type() else {
                continue;
            };
            if file_type.is_dir() {
                let name = entry.file_name().to_string_lossy().to_string();
                if !skip_dirs.contains(&name) {
                    subdirs.push(path);
                }
                continue;
            }
            if !file_type.is_file() {
                continue;
            }
            let name = entry.file_name().to_string_lossy().to_string();
            if let Some(dot) = name.rfind('.') {
                if dot > 0 {
                    let ext = name[dot + 1..].to_lowercase();
                    if video_exts.contains(&ext) {
                        results.push((path, dir.clone()));
                    }
                }
            }
        }
        stack.extend(subdirs.into_iter().rev());
    }
    results
}

fn display_base(filename: &str) -> String {
    match filename.rfind('.') {
        Some(0) | None => filename.to_string(),
        Some(idx) => filename[..idx].to_string(),
    }
}

fn make_unique_display_name(
    base: &str,
    source_rel: &str,
    used: &HashSet<String>,
) -> (String, bool) {
    if !used.contains(base) {
        return (base.to_string(), false);
    }
    let rel_suffix = source_rel.trim_matches('/').trim_start_matches("./");
    if !rel_suffix.is_empty() {
        let candidate = format!("{base} ({rel_suffix})");
        if !used.contains(&candidate) {
            return (candidate, true);
        }
    }
    let mut index = 2;
    loop {
        let candidate = format!("{base} ({index})");
        if !used.contains(&candidate) {
            return (candidate, true);
        }
        index += 1;
    }
}

/// 正式扫描结果：统计 + 待后台校验的 source_path 列表。
pub struct ScanPersistResult {
    pub stats: ScanStats,
    pub paths_to_validate: Vec<String>,
}

pub async fn scan_root(
    settings: &MediaSettings,
    songs: &SongRepository,
    root: &str,
    video_exts: &HashSet<String>,
    skip_dir_names: &HashSet<String>,
    options: ScanOptions,
) -> Result<ScanPersistResult, ScanError> {
    let root_path =
        std::fs::canonicalize(root).map_err(|_| ScanError::RootNotFound(root.to_string()))?;
    if !root_path.is_dir() {
        return Err(ScanError::RootNotFound(root.to_string()));
    }
    let root_str = root_path.to_string_lossy().to_string();

    let existing = songs.all_for_scan().await?;
    let mut existing_by_path: HashMap<String, i64> = existing
        .iter()
        .map(|s| (s.source_path.clone(), s.id))
        .collect();
    let mut name_to_path: HashMap<String, String> = existing
        .iter()
        .map(|s| (s.display_name.clone(), s.source_path.clone()))
        .collect();
    let mut used_names: HashSet<String> = name_to_path.keys().cloned().collect();

    let files = {
        let root_path = root_path.clone();
        let video_exts = video_exts.clone();
        let skip_dir_names = skip_dir_names.clone();
        tokio::task::spawn_blocking(move || {
            walk_video_files(&root_path, &video_exts, &skip_dir_names)
        })
        .await
        .unwrap_or_default()
    };
    info!("scan collect: root={root_str} files={}", files.len());

    // 预览且需要校验：先并发探测，再走决策逻辑。
    let mut playable_map: HashMap<String, bool> = HashMap::new();
    if options.validate && options.dry_run {
        let concurrency = settings.scan_validate_concurrency;
        let paths: Vec<String> = files
            .iter()
            .map(|(p, _)| p.to_string_lossy().to_string())
            .collect();
        let results: Vec<(String, bool)> = stream::iter(paths)
            .map(|path| {
                let settings = settings.clone();
                async move {
                    let ok = probe_video_playable(&settings, &path).await;
                    (path, ok)
                }
            })
            .buffer_unordered(concurrency)
            .collect()
            .await;
        playable_map.extend(results);
    }

    let mut stats = ScanStats::default();
    let mut to_create: Vec<NewSong> = Vec::new();
    let mut overwrite_updates: Vec<(i64, NewSong)> = Vec::new();
    let mut paths_to_validate: Vec<String> = Vec::new();

    for (abs_path, dirpath) in files {
        let abs_path_str = abs_path.to_string_lossy().to_string();
        let filename = abs_path
            .file_name()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string();
        let base_name = display_base(&filename);
        let mut source_rel = dirpath
            .strip_prefix(&root_path)
            .map(|p| p.to_string_lossy().to_string())
            .unwrap_or_default();
        if source_rel == "." {
            source_rel = String::new();
        }

        // 正式扫描：暂标可播，校验后台补齐；预览：同步用探测结果。
        let is_playable = if options.validate && options.dry_run {
            let ok = playable_map.get(&abs_path_str).copied().unwrap_or(false);
            if !ok {
                stats.invalid += 1;
                stats.push_preview(PreviewItem {
                    path: abs_path_str.clone(),
                    action: "invalid".to_string(),
                    display_name: None,
                });
                continue;
            }
            true
        } else {
            true
        };

        if let Some(&existing_id) = existing_by_path.get(&abs_path_str) {
            if options.duplicate_policy == "overwrite" {
                if options.dry_run {
                    stats.push_preview(PreviewItem {
                        path: abs_path_str.clone(),
                        action: "update".to_string(),
                        display_name: None,
                    });
                } else {
                    overwrite_updates.push((
                        existing_id,
                        NewSong {
                            display_name: base_name.clone(),
                            source_path: abs_path_str.clone(),
                            source_origin: "scan".to_string(),
                            source_rel: (!source_rel.is_empty()).then(|| source_rel.clone()),
                            media_kind: "video".to_string(),
                            playback_mode: "plain".to_string(),
                            playback_source: Some("plain".to_string()),
                            can_queue: Some(is_playable),
                            is_playable,
                            scan_root: Some(root_str.clone()),
                        },
                    ));
                    if options.validate {
                        paths_to_validate.push(abs_path_str.clone());
                    }
                }
                stats.added += 1;
            } else {
                stats.skipped += 1;
                if options.dry_run {
                    stats.push_preview(PreviewItem {
                        path: abs_path_str.clone(),
                        action: "skip".to_string(),
                        display_name: None,
                    });
                }
            }
            continue;
        }

        let mut display_name = base_name.clone();
        let mut renamed = false;
        if let Some(existing_path) = name_to_path.get(&display_name) {
            if existing_path != &abs_path_str {
                match options.duplicate_policy.as_str() {
                    "skip" => {
                        stats.skipped += 1;
                        if options.dry_run {
                            stats.push_preview(PreviewItem {
                                path: abs_path_str.clone(),
                                action: "skip".to_string(),
                                display_name: None,
                            });
                        }
                        continue;
                    }
                    "overwrite" => {
                        if !options.dry_run {
                            if let Some(&existing_id) = existing_by_path.get(existing_path) {
                                overwrite_updates.push((
                                    existing_id,
                                    NewSong {
                                        display_name: display_name.clone(),
                                        source_path: abs_path_str.clone(),
                                        source_origin: "scan".to_string(),
                                        source_rel: (!source_rel.is_empty())
                                            .then(|| source_rel.clone()),
                                        media_kind: "video".to_string(),
                                        playback_mode: "plain".to_string(),
                                        playback_source: Some("plain".to_string()),
                                        can_queue: Some(is_playable),
                                        is_playable,
                                        scan_root: Some(root_str.clone()),
                                    },
                                ));
                            }
                            existing_by_path.remove(existing_path);
                            existing_by_path.insert(abs_path_str.clone(), 0);
                            if options.validate {
                                paths_to_validate.push(abs_path_str.clone());
                            }
                        }
                        stats.added += 1;
                        if options.dry_run {
                            stats.push_preview(PreviewItem {
                                path: abs_path_str.clone(),
                                action: "overwrite".to_string(),
                                display_name: None,
                            });
                        }
                        continue;
                    }
                    _ => {
                        let (unique_name, was_renamed) =
                            make_unique_display_name(&base_name, &source_rel, &used_names);
                        display_name = unique_name;
                        renamed = was_renamed;
                        if renamed {
                            stats.renamed += 1;
                        }
                    }
                }
            }
        }

        used_names.insert(display_name.clone());
        name_to_path.insert(display_name.clone(), abs_path_str.clone());

        if options.dry_run {
            stats.push_preview(PreviewItem {
                path: abs_path_str.clone(),
                action: if renamed {
                    "rename".to_string()
                } else {
                    "add".to_string()
                },
                display_name: Some(display_name),
            });
            stats.added += 1;
            continue;
        }

        to_create.push(NewSong {
            display_name,
            source_path: abs_path_str.clone(),
            source_origin: "scan".to_string(),
            source_rel: (!source_rel.is_empty()).then_some(source_rel),
            media_kind: "video".to_string(),
            playback_mode: "plain".to_string(),
            playback_source: Some("plain".to_string()),
            can_queue: Some(is_playable),
            is_playable,
            scan_root: Some(root_str.clone()),
        });
        existing_by_path.insert(abs_path_str.clone(), 0);
        if options.validate {
            paths_to_validate.push(abs_path_str);
        }
        stats.added += 1;
    }

    if !options.dry_run {
        if !to_create.is_empty() {
            songs.bulk_insert(&to_create).await?;
        }
        for (id, update) in overwrite_updates {
            songs
                .apply_scan_overwrite(
                    id,
                    &update.source_path,
                    &update.display_name,
                    update.source_rel.as_deref(),
                    update.is_playable,
                    update.scan_root.as_deref().unwrap_or(&root_str),
                )
                .await?;
        }
        info!(
            "scan persist: root={root_str} created={} pending_validate={}",
            stats.added,
            paths_to_validate.len()
        );
    }

    stats.pending_validate = paths_to_validate.len() as i64;
    Ok(ScanPersistResult {
        stats,
        paths_to_validate,
    })
}

/// 后台并发校验：更新 is_playable / can_queue。
pub async fn run_validate_paths(
    settings: &MediaSettings,
    songs: &SongRepository,
    task: &ScanTaskManager,
    paths: Vec<String>,
) {
    if paths.is_empty() {
        return;
    }
    if !task.try_begin(paths.len() as i64) {
        tracing::warn!("scan validate skipped: another task is running");
        return;
    }

    let concurrency = settings.scan_validate_concurrency;
    stream::iter(paths)
        .map(|path| {
            let settings = settings.clone();
            let songs = songs.clone();
            let task = task.clone();
            async move {
                let ok = probe_video_playable(&settings, &path).await;
                if let Ok(Some(song)) = songs.find_by_source_path(&path).await {
                    if let Err(e) = songs.update_playable_flags(song.id, ok, ok).await {
                        tracing::warn!("update playable flags failed for {}: {e}", song.id);
                    }
                }
                task.tick(!ok);
            }
        })
        .buffer_unordered(concurrency)
        .collect::<Vec<_>>()
        .await;

    task.finish();
    info!("scan validate finished: {:?}", task.status());
}
